// @transtrack/ui/lib/commands/add.js
import chalk from "chalk";
import fs from "fs-extra";
import ora from "ora";
import path, { dirname } from "path";
import prompts from "prompts";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

import { installComponent } from '../utils/installer.js';

// Daftar component yang tersedia
const AVAILABLE_COMPONENTS = [
  'Avatar',
  'BorderedCheckboxGroup',
  'BorderdRadio',
  'Button',
  'ButtonBack',
  'ButtonGroup',
  'Calendar',
  'CalendarMonth',
  'CalendarWeek',
  'Checkbox',
  'Chip',
  'Collapse',
  'Combobox',
  'DataTable',
  'DateNavigation',
  'DatePicker',
  'DoughnutChart',
  'DragDropFile',
  'Dropdown',
  'FileButton',
  'IconButton',
  'Image',
  'ItemGroup',
  'LeafletMap',
  'LineChart',
  'Link',
  'Modal',
  'ModalConfirmation',
  'ModalLoading',
  'ModalWarning',
  'Notification',
  'NumberInput',
  'OutlinedButton',
  'Pagination',
  'Radio',
  'RadioGroup',
  'RangeDatePicker',
  'SpinnerNumberInput',
  'Stepper',
  'Tab',
  'TextArea',
  'TextButton',
  'TextInput',
  'Timeline',
  'Toggle',
  'Tooltip',
  'UploadClickDrop'
];

// Daftar icon yang tersedia
const AVAILABLE_ICONS = [
  'Alert',
  'AlertTriangle',
  'AnnotationQuestion',
  'ArrowLeft',
  'ArrowRight',
  'BarChart',
  'Bell',
  'Box',
  'CalendarIcon',
  'CalendarDate',
  'CameraOff',
  'Check',
  'CheckVerified',
  'ChevronDown',
  'ChevronLeft',
  'ChevronRight',
  'ChevronUp',
  'Circle',
  'Close',
  'Cube',
  'DownloadCloud',
  'Edit',
  'Expand',
  'Eye',
  'EyeClosed',
  'EyeOff',
  'File',
  'FileSearch',
  'FileText',
  'FilterLines',
  'Folder',
  'Info',
  'Key',
  'Url',
  'UrlBroken',
  'List',
  'Loading',
  'Logout',
  'MailOutline',
  'MapOutline',
  'Marker',
  'MarkerIdle',
  'MarkerOffline',
  'MarkerOnline',
  'MarkerPin',
  'Menu',
  'Minus',
  'Package',
  'PhoneOutline',
  'PieChart',
  'Plus',
  'Refresh',
  'RefreshDouble',
  'Repeat',
  'Route',
  'Search',
  'Settings',
  'Speedometer',
  'Trash',
  'UploadCloud',
  'User',
  'UsersPlus',
  'VersionControl'
];

// Fungsi untuk menentukan tipe berdasarkan nama
function determineType(name) {
  if (AVAILABLE_ICONS.includes(name)) {
    return 'icons';
  }
  if (AVAILABLE_COMPONENTS.includes(name)) {
    return 'components';
  }
  return null;
}

export default async function(components, options) {
  // Check if package.json exists and transtrack-ui is installed
  let packageJson = JSON.parse(
    fs.readFileSync(path.resolve(process.cwd(), "package.json"), "utf8")
  );
  


  if (!packageJson.dependencies?.["@transtrack/ui"]) {
      console.log(chalk.red('\ntranstrack-ui belum terinstal. Silakan install terlebih dahulu:'));
    console.log(chalk.cyan('\nnpm install transtrack-ui'));
    return;
  }

  // Jika tidak ada component yang ditentukan, tampilkan UI untuk memilih
  if (!components || components.length === 0) {
    const allItems = [...AVAILABLE_COMPONENTS, ...AVAILABLE_ICONS];
    const response = await prompts({
      type: 'multiselect',
      name: 'components',
      message: 'Pilih component atau icon yang akan diinstal',
      choices: allItems.map(c => ({ 
        title: c,
        value: c,
        description: AVAILABLE_ICONS.includes(c) ? 'Icon' : 'Component'
      })),
      hint: '- Spasi untuk memilih. Enter untuk konfirmasi'
    });
    
    components = response.components;
    
    if (!components || components.length === 0) {
      console.log(chalk.yellow('\nTidak ada yang dipilih. Keluar.'));
      return;
    }
  }
  
  // Kelompokkan component berdasarkan tipe
  const groupedItems = components.reduce((acc, item) => {
    const type = determineType(item);
    if (!type) {
      acc.invalid.push(item);
    } else {
      if (!acc[type]) acc[type] = [];
      acc[type].push(item);
    }
    return acc;
  }, { invalid: [] });

  // Tampilkan error untuk item yang tidak valid
  if (groupedItems.invalid.length > 0) {
    console.log(chalk.red('\nComponent tidak tersedia, pastikan nama component benar dan check dokumentasi.'));
    // TODO: ADD URL STORY BOOK TRANSTRACK UI
    // console.log(groupedItems.invalid.map(c => `  ${c}`).join('\n'));
    // console.log(chalk.cyan('\nComponent yang tersedia:'));
    // console.log(AVAILABLE_COMPONENTS.map(c => `  ${c}`).join('\n'));
    // console.log(chalk.cyan('\nIcon yang tersedia:'));
    // console.log(AVAILABLE_ICONS.map(c => `  ${c}`).join('\n'));
    return;
  }

  // Instal setiap item berdasarkan tipenya
  for (const [type, items] of Object.entries(groupedItems)) {
    if (items.length === 0) continue;

    const targetDir = type === 'icons' ? 'components/icons' : options.directory;
    fs.ensureDirSync(targetDir);

    for (const item of items) {
      const spinner = ora(`Menginstal ${type === 'icons' ? 'icon' : 'component'} `).start();
      
      try {
        spinner.stop();
        const result = await installComponent(item, targetDir, type);
        if (result === false) {
          // Jika instalasi dibatalkan, tidak perlu menampilkan pesan sukses
          continue;
        }
        spinner.succeed(`${item} berhasil ditambahkan.`);
      } catch (err) {
        spinner.fail(`Gagal menginstal ${item}: ${err.message}`);
        console.error(err);
      }
    }
  }
}
