// @transtrack/ui/lib/utils/installer.js
import chalk from 'chalk';
import fs from 'fs-extra';
import path, { dirname } from 'path';
import prompts from 'prompts';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * Mendapatkan path ke sumber component/icon
 * @param {string} name - Nama component/icon
 * @param {string} type - Tipe (components/icons)
 * @returns {string} - Path ke direktori sumber
 */
export function getComponentSourcePath(name, type = 'components') {
  const baseDir = type === 'icons' ? 'icons' : 'general';
  return path.resolve(__dirname, '../../components', baseDir, name);
}

/**
 * Menginstal component/icon ke proyek pengguna
 * @param {string} name - Nama component/icon yang akan diinstal
 * @param {string} targetDir - Direktori tujuan
 * @param {string} type - Tipe (components/icons)
 * @returns {Promise<boolean>} - Hasil instalasi
 */
export async function installComponent(name, targetDir, type = 'components') {
  const sourcePath = getComponentSourcePath(name, type);
  const destPath = path.resolve(process.cwd(), targetDir, name);
  
  // Pastikan sumber ada
  if (!fs.existsSync(sourcePath)) {
    throw new Error(`${type === 'icons' ? 'Icon' : 'Component'} ${name} tidak ditemukan`);
  }

  // Cek apakah component sudah ada
  if (fs.existsSync(destPath)) {
    const response = await prompts({
      type: 'confirm',
      name: 'overwrite',
      message: chalk.red(`${type === 'icons' ? 'Icon' : 'Component'} ${name} sudah ada. Apakah Anda ingin menimpanya?`),
      initial: false
    });

    if (!response.overwrite) {
      console.log(chalk.blue(`\nInstalasi ${type === 'icons' ? 'icon' : 'component'} ${name} dibatalkan.`));
      return false;
    }
  }

  // Salin file index.vue utama
  const mainVueFile = path.join(sourcePath, 'index.vue');
  if (fs.existsSync(mainVueFile)) {
    await fs.ensureDir(destPath);
    await fs.copy(mainVueFile, path.join(destPath, 'index.vue'));
  } else {
    throw new Error(`File  tidak ditemukan.`);
  }
  
  return true;
}
